code_folder/22_glue_s3_athena_pipeline.py

```
@@ -0,0 +1,617 @@
+        print(f"[Export] Wrote {written} items to DynamoDB table '{table_name}'")
+
+    @staticmethod
+    def to_s3_for_redshift(df: pd.DataFrame, bucket: str, key: str) -> str:
+        """Write CSV to S3 for Redshift COPY command."""
+        import boto3
+        s3 = boto3.client("s3")
+        csv_data = df.to_csv(index=False)
+        s3.put_object(Bucket=bucket, Key=key, Body=csv_data.encode())
+        s3_uri = f"s3://{bucket}/{key}"
+        print(f"[Export] Wrote {len(df)} rows to {s3_uri} (ready for Redshift COPY)")
+        print(f"  -> COPY command:")
```

Ruff (0.14.14)

[error] f-string without any placeholders

Remove extraneous `f` prefix

(F541)